datafusion/physical-expr/src/equivalence.rs

```diff
@@ -20,26 +20,114 @@ use std::hash::Hash;
 use std::sync::Arc;
 
 use crate::expressions::Column;
-use crate::physical_expr::{deduplicate_physical_exprs, have_common_entries};
 use crate::sort_properties::{ExprOrdering, SortProperties};
 use crate::{
-    physical_exprs_contains, LexOrdering, LexOrderingRef, LexRequirement,
-    LexRequirementRef, PhysicalExpr, PhysicalSortExpr, PhysicalSortRequirement,
+    physical_exprs_bag_equal, physical_exprs_contains, physical_exprs_equal, LexOrdering,
+    LexOrderingRef, LexRequirement, LexRequirementRef, PhysicalExpr, PhysicalSortExpr,
+    PhysicalSortRequirement,
 };
 
 use arrow::datatypes::SchemaRef;
 use arrow_schema::SortOptions;
 use datafusion_common::tree_node::{Transformed, TreeNode};
 use datafusion_common::{JoinSide, JoinType, Result};
 
+use crate::physical_expr::deduplicate_physical_exprs;
 use indexmap::map::Entry;
 use indexmap::IndexMap;
 
 /// An `EquivalenceClass` is a set of [`Arc<dyn PhysicalExpr>`]s that are known
 /// to have the same value for all tuples in a relation. These are generated by
-/// equality predicates, typically equi-join conditions and equality conditions
-/// in filters.
-pub type EquivalenceClass = Vec<Arc<dyn PhysicalExpr>>;
+/// equality predicates (e.g. `a = b`), typically equi-join conditions and
+/// equality conditions in filters.
+#[derive(Debug, Clone)]
+pub struct EquivalenceClass {
```

Comment:
@ozankabak  and @mustafasrepo  I am curious what you think about this change to encapsulate free functions into a struct in the equivalence code? If you like it, I can pull most of the rest of the code in datafusion/physical-expr/src/physical_expr.rs to this structure instead of free functions

Reply:
I like this style, `EquivalenceClass` should be its own type. When comparing two equivalence classes, calling `eq` on this type is much more readable than calling a utility that operates on `PhysicalExpr` slices directly.

You seem to be keeping the lower-level functional primitives around, which is also good -- we can reuse them to create encapsulations like `EquivalenceClass` as we build more.

Reply:
> You seem to be keeping the lower-level functional primitives around, which is also good -- we can reuse them to create encapsulations like EquivalenceClass as we build more.

I did this mostly to keep the size of the initial diff down to make the proposal easier to review.

It seems to me like `EquivalenceClass` has very few functions actually related to equivalence calculations -- it is mostly a container of `PhysicalExpr`s -- maybe it would be better named something like `PhysicalExprList` 🤔 But then the equivalence calculations would be less readable perhaps.

Reply:
Exactly. I see two options:
1. Following your approach where we encapsulate free functions in structs and methods whenever we can define a coherent struct for them, and we also keep these free functions so that similar structs can share them.
2. Create agnostic structs (like `PhysicalExprList`) and convert the free functions to their methods, and use type aliases like `EquivalenceClass` = `PhysicalExprList` to increase readability depending on the context.

Reply:
I would like to propose with starting with EquivalenceClass into its own struct and we can keep the utility functions around for a while to see if they are reusable